File: implementations/java/vanilla/src/main/java/org/cwi/waebric/parser/ast/markup/Markup.java
package org.cwi.waebric.parser.ast.markup;

import org.cwi.waebric.parser.ast.SyntaxNode;
import org.cwi.waebric.parser.ast.INodeVisitor;

/**
 * Markup
 * @author Jeroen van Schagen
 * @date 22-05-2009
 */
public abstract class Markup extends SyntaxNode {
	
	protected Designator designator;
	
	/**
	 * Retrieve designator
	 * @return
	 */
	public Designator getDesignator() {
		return designator;
	}

	/**
	 * Store designator
	 * @param designator
	 */
	public void setDesignator(Designator designator) {
		this.designator = designator;
	}

	/**
	 * Designator Arguments -> Markup
	 * @author Jeroen van Schagen
	 * @date 20-05-2009
	 */
	public static class Call extends Markup {

		protected Arguments arguments;

		public Call(Designator designator) {
			this(designator, new Arguments());
		}

		public Call(Designator designator, Arguments arguments) {
			this.designator = designator;
			this.arguments = arguments;
		}
		
		/**
		 * Retrieve arguments
		 * @return
		 */
		public Arguments getArguments() {
			return arguments;
		}
		
		public void setArguments(Arguments arguments) {
			this.arguments = arguments;
		}
		
		public SyntaxNode[] getChildren() {
			return new SyntaxNode[] { designator, arguments };
		}
		
			@Override
	public <T> T accept(INodeVisitor<T> visitor) {
		return visitor.visit(this);
	}
		
	}
	
	/**
	 * Designator -> Markup
	 * @author Jeroen van Schagen
	 * @date 20-05-2009
	 */
	public static class Tag extends Markup {

		public Tag(Designator designator) {
			this.designator = designator;
		}
		
		public SyntaxNode[] getChildren() {
			return new SyntaxNode[] { designator };
		}
		
			@Override
	public <T> T accept(INodeVisitor<T> visitor) {
		return visitor.visit(this);
	}
		
	}

}

File: implementations/csharp/vanilla/parser/ast/Markup/Markup.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Parser.Ast.Markup
{
    /// <summary>
    /// Node containing Markup
    /// </summary>
    public class Markup : ISyntaxNode
    {
        #region Private Members

        private NodeList ArgumentList;
        private Designator Tag;
        private bool IsCall;

        #endregion

        #region Public Methods

        public Markup()
        {
            //Intialize members
            ArgumentList = new NodeList();
            IsCall = false;
        }

        /// <summary>
        /// Set Designator tag
        /// </summary>
        /// <param name="tag">Tag to set</param>
        public void SetDesignator(Designator tag)
        {
            Tag = tag;
        }

        /// <summary>
        /// Get Designator tag
        /// </summary>
        /// <returns>Tag</returns>
        public Designator GetDesignator()
        {
            return Tag;
        }

        /// <summary>
        /// Add argument to markup
        /// </summary>
        /// <param name="argument">Argument to add</param>
        public void AddArgument(Argument argument)
        {
            ArgumentList.Add(argument);
        }

        /// <summary>
        /// Get arguments
        /// </summary>
        /// <returns>ArgumentList</returns>
        public NodeList GetArguments()
        {
            return ArgumentList;
        }

        /// <summary>
        /// Set Markup as call
        /// </summary>
        public void SetCall()
        {
            IsCall = true;
        }

        /// <summary>
        /// Get state of Call
        /// </summary>
        /// <returns></returns>
        public bool GetCallState()
        {
            return IsCall;
        }

        public override String ToString()
        {
            if(ArgumentList.Count == 0)
            {
                return Tag.ToString();
            }
            else
            {
                //Get arguments
                String buffer = "";
                int counter = 0;
                foreach (Argument node in ArgumentList)
                {
                    buffer += node.ToString();
                    if (counter != (ArgumentList.Count - 1))
                    {
                        buffer += ",";
                    }
                    counter++;
                }
                return Tag.ToString() + "(" + buffer + ")";
            }
        }

        public void AcceptVisitor(ISyntaxNodeVisitor visitor)
        {
            visitor.Visit(this);
        }

        public ISyntaxNode[] GetSubNodes()
        {
            return new ISyntaxNode[] { 
                Tag,
                ArgumentList
            };
        }

        #endregion

    }
}
